candidate.go

```diff
@@ -20,260 +16,26 @@ const (
-			log.Warnf("Discarded message from %s, not a valid remote candidate", c.addr())
-		}
-
-		// NOTE This will return packetio.ErrFull if the buffer ever manages to fill up.
-		_, err = c.agent.buffer.Write(buffer[:n])
-		if err != nil {
-			log.Warnf("failed to write packet")
-		}
-	}
-}
-
-// close stops the recvLoop
-func (c *Candidate) close() error {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-	if c.conn != nil {
-		// Unblock recvLoop
-		close(c.closeCh)
-		// Close the conn
-		err := c.conn.Close()
-		if err != nil {
-			return err
-		}
-
-		// Wait until the recvLoop is closed
-		<-c.closedCh
-	}
-
-	return nil
-}
-
-func (c *Candidate) writeTo(raw []byte, dst *Candidate) (int, error) {
-	n, err := c.conn.WriteTo(raw, dst.addr())
-	if err != nil {
-		return n, fmt.Errorf("failed to send packet: %v", err)
-	}
-	c.seen(true)
-	return n, nil
-}
-
-// Priority computes the priority for this ICE Candidate
-func (c *Candidate) Priority() uint32 {
-	// The local preference MUST be an integer from 0 (lowest preference) to
-	// 65535 (highest preference) inclusive.  When there is only a single IP
-	// address, this value SHOULD be set to 65535.  If there are multiple
-	// candidates for a particular component for a particular data stream
-	// that have the same type, the local preference MUST be unique for each
-	// one.
-	return (1<<24)*uint32(c.Type.Preference()) +
-		(1<<8)*uint32(c.LocalPreference) +
-		uint32(256-c.Component)
-}
-
-// Equal is used to compare two CandidateBases
-func (c *Candidate) Equal(other *Candidate) bool {
-	return c.NetworkType == other.NetworkType &&
-		c.Type == other.Type &&
-		c.IP.Equal(other.IP) &&
-		c.Port == other.Port &&
-		c.RelatedAddress.Equal(other.RelatedAddress)
-}
-
-// String makes the CandidateHost printable
-func (c *Candidate) String() string {
-	return fmt.Sprintf("%s %s:%d%s", c.Type, c.IP, c.Port, c.RelatedAddress)
-}
-
-// LastReceived returns a time.Time indicating the last time
-// this candidate was received
-func (c *Candidate) LastReceived() time.Time {
-	c.lock.RLock()
-	defer c.lock.RUnlock()
-	return c.lastReceived
-}
-
-func (c *Candidate) setLastReceived(t time.Time) {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-	c.lastReceived = t
-}
-
-// LastSent returns a time.Time indicating the last time
-// this candidate was sent
-func (c *Candidate) LastSent() time.Time {
-	c.lock.RLock()
-	defer c.lock.RUnlock()
-	return c.lastSent
-}
-
-func (c *Candidate) setLastSent(t time.Time) {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-	c.lastSent = t
-}
-
-func (c *Candidate) seen(outbound bool) {
-	if outbound {
-		c.setLastSent(time.Now())
-	} else {
-		c.setLastReceived(time.Now())
-	}
-}
-
-func (c *Candidate) addr() net.Addr {
-	return &net.UDPAddr{
-		IP:   c.IP,
-		Port: c.Port,
-	}
+type Candidate interface {
+	start(a *Agent, conn net.PacketConn)
+	addr() net.Addr
+
+	setLastSent(t time.Time)
+	seen(outbound bool)
+	LastSent() time.Time
+	setLastReceived(t time.Time)
```

Comment:
probably you can remove this from the interface

Reply:
Fixed!